File: packages/server/src/routes/orchestrate.ts
/**
 * Orchestration (Airflow, Dagster, Prefect, cron, CI):
 *   POST /api/orchestrate/runs {kind, id, …, wait?}        start a run (write scope); wait: true answers when it is done
 *   GET  /api/orchestrate/runs/:id?wait=30                 its status: running | succeeded | failed (long-polls up to 60 s)
 *   GET  /api/orchestrate/runs                             your recent runs
 */
import type { FastifyInstance } from 'fastify';
import { z } from 'zod';
import type { AppContext } from '../context.js';
import { ORCHESTRATION_KINDS } from '../db/schema/sqlite.js';

const Body = z.object({
  kind: z.enum(ORCHESTRATION_KINDS),
  id: z.string().min(1).max(64),
  command: z.enum(['build', 'run', 'test', 'seed', 'compile']).nullable().optional(),
  select: z.string().max(2000).nullable().optional(),
  exclude: z.string().max(2000).nullable().optional(),
  full_refresh: z.boolean().optional(),
  sql: z.string().max(100_000).nullable().optional(),
  fail_if: z.enum(['rows', 'no_rows']).nullable().optional(),
  input: z.string().max(8000).nullable().optional(),
  fail_on_warn: z.boolean().optional(),
  fail_on_trigger: z.boolean().optional(),
  fail_on_anomaly: z.boolean().optional(),
  source: z.string().max(40).nullable().optional(),
  external_run_id: z.string().max(250).nullable().optional(),
  wait: z.boolean().optional(),
});

export async function orchestrateRoutes(app: FastifyInstance, ctx: AppContext) {
  app.addHook('preHandler', app.authenticate);
  app.post('/api/orchestrate/runs', async (req, reply) => {
    const { wait, ...input } = Body.parse(req.body ?? {});
    const run = await ctx.orchestrate.start(req.principal!, input, { wait });
    return reply.code(wait ? 200 : 202).send({ run });
  });
  app.get('/api/orchestrate/runs/:id', async (req) => {
    const q = z.object({ wait: z.coerce.number().min(0).max(60).optional() }).parse(req.query ?? {});
    const { id } = req.params as { id: string };
    return { run: q.wait ? await ctx.orchestrate.wait(req.principal!, id, q.wait) : await ctx.orchestrate.get(req.principal!, id) };
  });
  app.get('/api/orchestrate/runs', async (req) => {
    const q = z.object({ limit: z.coerce.number().int().min(1).max(500).optional() }).parse(req.query ?? {});
    return { runs: await ctx.orchestrate.list(req.principal!, q.limit) };
  });
}

File: packages/sdk-python/duckview/orchestrate.py
"""
Run DuckView work from an orchestrator (Airflow, Dagster, Prefect, cron, CI) and wait for one status.

    from duckview.orchestrate import run
    run("sync", "<sync id>")                                   # load a sync; raises RunFailed if it fails
    run("dbt", "<project id>", command="build", select="tag:daily")
    run("quality", "<suite id>", fail_on_warn=True)
    run("query", "<workspace id>", sql="SELECT * FROM orders WHERE amount < 0", fail_if="rows")

Kinds: sync, dbt, quality, reverse_sync, notebook, alert, snapshot, agent, monitor, query. The run starts in
DuckView (POST /api/orchestrate/runs) and is long-polled until it has finished; its record (status, summary,
detail) is returned. The token needs the write scope; runs act as the token's owner.
"""
from __future__ import annotations

import time
from typing import Any, Dict, Optional

from . import Client, DuckViewError, connect

__all__ = ["run", "start", "wait", "RunFailed", "KINDS"]

KINDS = ("sync", "dbt", "quality", "reverse_sync", "notebook", "alert", "snapshot", "agent", "monitor", "query")


class RunFailed(Exception):
    """A DuckView run that finished as failed (a failing check, an error)."""

    def __init__(self, run: Dict[str, Any]):
        super().__init__(f"DuckView {run.get('kind')} '{run.get('label')}' failed: {run.get('summary')}")
        self.run = run


def start(kind: str, target: str, client: Optional[Client] = None, **options: Any) -> Dict[str, Any]:
    """Starts a run and returns it at once (status "running")."""
    if kind not in KINDS:
        raise ValueError(f"kind must be one of {', '.join(KINDS)}")
    c = client or connect()
    body = {k: v for k, v in options.items() if v is not None}
    body.update({"kind": kind, "id": target})
    return c._request("POST", "/api/orchestrate/runs", body)["run"]


def wait(run_id: str, client: Optional[Client] = None, timeout: Optional[float] = None, poll: float = 30.0) -> Dict[str, Any]:
    """Long-polls a run until it is no longer running (or `timeout` seconds pass: TimeoutError)."""
    c = client or connect()
    deadline = None if timeout is None else time.monotonic() + timeout
    while True:
        step = poll if deadline is None else max(0.0, min(poll, deadline - time.monotonic()))
        r = c._request("GET", f"/api/orchestrate/runs/{run_id}?wait={int(max(1, min(60, step)))}")["run"]
        if r["status"] != "running":
            return r
        if deadline is not None and time.monotonic() >= deadline:
            raise TimeoutError(f"DuckView run {run_id} still running after {timeout} s")


def run(kind: str, target: str, client: Optional[Client] = None, timeout: Optional[float] = None, raise_on_failure: bool = True, **options: Any) -> Dict[str, Any]:
    """Starts a run, waits for it, and raises RunFailed when it failed (unless raise_on_failure=False)."""
    c = client or connect()
    r = wait(start(kind, target, client=c, **options)["id"], client=c, timeout=timeout)
    if r["status"] == "failed" and raise_on_failure:
        raise RunFailed(r)
    return r
